state/protocol/badger/state.go

```diff
@@ -82,23 +82,30 @@ func Bootstrap(
 		return nil, fmt.Errorf("cannot bootstrap invalid root snapshot: %w", err)
 	}
 
-	err = operation.RetryOnConflictTx(db, transaction.Update, func(tx *transaction.Tx) error {
+	segment, err := root.SealingSegment()
+	if err != nil {
+		return nil, fmt.Errorf("could not get sealing segment: %w", err)
+	}
 
-		// 1) insert each block in the root chain segment
-		err = state.bootstrapSealingSegment(root)(tx)
-		if err != nil {
-			return fmt.Errorf("could not bootstrap sealing chain segment: %w", err)
-		}
+	// bootstrap the sealing segment step 1: insert segment execution results.
+	// execution results must be inserted first.
+	err = state.bootstrapSealingSegmentExeResults(segment)
```

Comment:
Any reason this is not part of the database transaction like the subsequent state modifications? (ie. within `RetryOnConflictTx` and using `transaction.Tx`). It should be, otherwise it is possible to only partially bootstrap the state if an error occurs.

Reply:
Transactions are executed async so there is still the case of receipts getting stored without results. Thats why results are stored in their own transaction, before receipts transaction.

Reply:
Transactions _are_ executed asynchronously, but storage operations are blocking, so each operation within a transaction will occur strictly consecutively, unless the transaction user is introducing concurrency. 

If I have a transaction like:
```go
db.Update(func(*badger.Txn) error {
   operation.InsertSomething(A)(tx)
   operation.InsertSomething(B)(tx)
}
```
I know that `A` is inserted when `B`.

From Badger's [blog post about this](https://dgraph.io/blog/post/badger-txn/):
> As writes are performed in the transaction, we maintain an internal cache of these writes within the transaction. Thus, any follow up reads on the same key by this transaction would see this write. But, other transactions won’t, providing Isolation and Consistency.

Reply:
+1 on this, better to be consistent and use tx as well